### ISSUE TITLE
why not use batched_nms to replace agnostic ? thanks so much !

### ISSUE THREAD

**Issue by Yaoxingtian** [ultralytics/yolov5#5261]

**Comment by glenn-jocher** [comment]
@Yaoxingtian batched NMS is used by default in YOLOv5. It has no relation to agnostic, which is an NMS option.

**Comment by eoinoconn** [comment]
@glenn-jocher I think what @Yaoxingtian is asking is why do you perform batched NMS by shifting all boxes by their class index and max_wh and calling torchvision.ops.nms(). Surely using torchvision.ops.batched_nms() would be faster as it wouldn't attempt to compare boxes that are not the same class?

https://github.com/ultralytics/yolov5/blob/19c8760caa70f4d04d3fb974d797f9d922bf6eb8/utils/general.py#L689-L691

https://pytorch.org/vision/stable/ops.html#torchvision.ops.batched_nms

**Comment by glenn-jocher** [comment]
@eoinoconn ah I see. I think the process is similar internally with batched_nms, but the last time we profiled this option was slower than the current. If this has changed please let us know though, we'd love to be able to speed this up if possible.

**Comment by glenn-jocher** [comment]
@Yaoxingtian @eoinoconn line-profiling comparison from today shows `ops.nms` is faster than `ops.batched_nms`. This is consistent with past results.

<img width="859" alt="Screenshot 2021-11-02 at 23 32 57" src="https://user-images.githubusercontent.com/26833433/139961099-533b29e0-e921-410b-b7f7-db85a4869c44.png">

**Comment by Yaoxingtian** [comment]
> @glenn-jocher I think what @Yaoxingtian is asking is why do you perform batched NMS by shifting all boxes by their class index and max_wh and calling torchvision.ops.nms(). Surely using torchvision.ops.batched_nms() would be faster as it wouldn't attempt to compare boxes that are not the same class?
> 
> https://github.com/ultralytics/yolov5/blob/19c8760caa70f4d04d3fb974d797f9d922bf6eb8/utils/general.py#L689-L691
> 
> https://pytorch.org/vision/stable/ops.html#torchvision.ops.batched_nms

Yes, this is what I want to  say. Thanks so much!

**Comment by Yaoxingtian** [comment]
> @Yaoxingtian @eoinoconn line-profiling comparison from today shows `ops.nms` is faster than `ops.batched_nms`. This is consistent with past results.
> 
> <img alt="Screenshot 2021-11-02 at 23 32 57" width="859" src="https://user-images.githubusercontent.com/26833433/139961099-533b29e0-e921-410b-b7f7-db85a4869c44.png">

@glenn-jocher @eoinoconn thanks for your reply ! I've tried to use agnostic and torchvision.ops.batched_nms, and they are both help to eliminate overlap boxes. so can agnostic replaces by batched_nms?  @glenn-jocher I see the result you showed, could you tell me what the tool you use to calculate the op time ?

**Comment by zhiqwang** [comment]
FYI, Previously, in order to export the post-processing part to onnx and torchscript graph, I've implemented a `batched_nms` version of YOLOv5, the current code structure will be a little more streamlined, as the price, will reduce part of the functionality. At the moment I can [guarantee](https://zhiqwang.com/yolov5-rt-stack/notebooks/how-to-align-with-ultralytics-yolov5.html#Varify-the-detection-results-between-yolort-and-ultralytics) that the accuracy of the two is basically the same, but I have not done a careful comparison in terms of inference time. I guess that this implementation can be used as a reference.

https://github.com/zhiqwang/yolov5-rt-stack/blob/284e3e2/yolort/models/box_head.py#L371-L385

```python
# Compute conf
# box_conf x class_conf, w/ shape: num_anchors x num_classes
scores = pred_logits[:, 5:] * pred_logits[:, 4:5]

boxes = det_utils.decode_single(pred_logits[:, :4], anchors_tuple)

# remove low scoring boxes
inds, labels = torch.where(scores > self.score_thresh)
boxes, scores = boxes[inds], scores[inds, labels]

# non-maximum suppression, independently done per level
keep = box_ops.batched_nms(boxes, scores, labels, self.nms_thresh)
# keep only topk scoring head_outputs
keep = keep[: self.detections_per_img]
boxes, scores, labels = boxes[keep], scores[keep], labels[keep]
```

**Comment by glenn-jocher** [comment]
@Yaoxingtian as I said before agnostic has nothing to do with batched vs non-batched, which are just methods of implementing NMS. Agnostic is a setting that allows you to apply NMS treating all boxes as class-agnostic.

Profiling tool is line profiler.

**Comment by glenn-jocher** [comment]
Bench script for NMS:

```python
        from utils.torch_utils import time_sync

        # Batched NMS
        c = x[:, 5] * (0 if agnostic else 1)  # classes
        boxes, scores = x[:, :4], x[:, 4]  # boxes (offset by class), scores
        t = time_sync()
        for _ in range(100):
            i = torchvision.ops.batched_nms(boxes, scores, c, iou_thres)  # NMS
        print('torchvision.ops.batched_nms', time_sync() - t)

        # NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        t = time_sync()
        for _ in range(100):
            j = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        print('torchvision.ops.nms', time_sync() - t)

        assert torch.allclose(i, j), 'allclose failure'
```
